core/src/main/java/org/pentaho/di/connections/vfs/BaseVFSConnectionProvider.java

```diff
@@ -90,4 +101,65 @@ protected static boolean getBooleanValueOfVariable( VariableSpace space, String
   protected VariableSpace getSpace( ConnectionDetails connectionDetails ) {
     return connectionDetails.getSpace() == null ? Variables.getADefaultVariableSpace() : connectionDetails.getSpace();
   }
+
+  @Override
+  public boolean test( T connectionDetails, VFSConnectionTestOptions connectionTestOptions ) throws KettleException {
+    boolean valid = test( connectionDetails );
+    if ( !valid ) {
+      return false;
+    }
+
+    if ( !connectionDetails.isSupportsRootPath()  || connectionTestOptions.isIgnoreRootPath() ) {
+      return true;
+    }
+
+    String resolvedRootPath = getResolvedRootPath( connectionDetails );
+    if ( StringUtils.isEmpty( resolvedRootPath ) ) {
+      return !connectionDetails.isRootPathRequired();
+    }
+
+    String internalUrl = buildUrl( connectionDetails, resolvedRootPath );
+    FileObject fileObject = KettleVFS.getFileObject( internalUrl, new Variables(), getOpts( connectionDetails ) );
+
+    try {
+      return fileObject.exists();
+    } catch ( FileSystemException fileSystemException ) {
+      LOGGER.error( fileSystemException.getMessage() );
+      return false;
+    }
+  }
+
+  @Override
+  public String getResolvedRootPath( T connectionDetails ) {
+    if ( StringUtils.isNotEmpty( connectionDetails.getRootPath() ) ) {
+      VariableSpace space = getSpace( connectionDetails );
+      String resolvedRootPath = getVar( connectionDetails.getRootPath(), space );
+      if ( StringUtils.isNotBlank( resolvedRootPath ) ) {
+        return normalizeRootPath( resolvedRootPath );
+      }
+    }
+
+    return StringUtils.EMPTY;
+  }
+
+  private String normalizeRootPath( String rootPath ) {
+    if ( StringUtils.isNotEmpty( rootPath ) ) {
+      if ( !rootPath.startsWith( DELIMITER ) ) {
+        rootPath = DELIMITER + rootPath;
+      }
+      if (rootPath.endsWith( DELIMITER ) ) {
+        rootPath = rootPath.substring( 0, rootPath.length() - 1 );
+      }
+    }
+    return rootPath;
+  }
+
+  private String buildUrl( VFSConnectionDetails connectionDetails, String rootPath ) {
+    String domain = connectionDetails.getDomain();
+    if ( !domain.isEmpty() ) {
+      domain = DELIMITER + domain;
+    }
+    return connectionDetails.getType() + ":/" + domain + rootPath;
```

Comment:
Let's say the final absolute VFS URI for S3 in this "test" operation scenario is:
`s3://someBucket/superSecretFolder/someDirA/SomeDirB/SomeDirC/someRandomFilet.txt`

is the "root path" concept:
1. option 1: "absolute URI prefix" ie `s3://someBucket/superSecretFolder/someDirA/`
2. option2: "relative path prefix" ie `superSecretFolder/someDirA/SomeDirB/SomeDirC` (the assumption here is that the bucket would be in domain string)

If i read the code correct, BaseVFSConnectionProvider#buildUrl is implemented as option 2 and that goes against the current specs of the jira issue - https://hv-eng.atlassian.net/browse/BACKLOG-39677

Reply:
AFAIK, in this case, the root path would be `/someBucket/superSecretFolder/someDirA`.
The domain is not used when connections have buckets: they are mutually exclusive.

You can argue that the first segment of the URL should always coincide with the domain. But that's not how these things were designed initially.

After root path is developed, it might even be possible to get rid of domain in favor of root path, at least for any new developments.

Reply:
I agree having the user know our internal schema prefix `"s3"` for aws S3 or `"abfss"` for Azure Gen2 is a burden/tedious.

But from my understanding the PR code is option 1 , and jira "design specs" is option 2, no?

Reply:
The code implements neither of these options. JIRA is option 1, needs to be fixed.

Reply:
I will update the JIRA issue to reflect what is being implemented.
Also, I should mention that the fact that the root path would not include the protocol section was discussed in a team meeting.

Reply:
see comment : https://github.com/pentaho/pentaho-kettle/pull/9305/files#r1592967128